README.md
# wanderlust

FIXME: description

## Installation

Download from http://example.com/FIXME.

## Usage

FIXME: explanation

    $ java -jar wanderlust-0.1.0-standalone.jar [args]

## Options

FIXME: listing of options this app accepts.

## Examples

...

### Bugs

...

### Any Other Sections
### That You Think
### Might be Useful

## License

Copyright © 2016 FIXME

Distributed under the Eclipse Public License either version 1.0 or (at
your option) any later version.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased][unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2016-10-31
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2016-10-31
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[unreleased]: https://github.com/your-name/wanderlust/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/wanderlust/compare/0.1.0...0.1.1


test/wanderlust/core_test.clj
(ns wanderlust.core-test
  (:require [clojure.test :refer :all]
            [wanderlust.core :refer :all]))

(deftest a-test
  (testing "FIXME, I fail."
    (is (= 0 1))))


src/wanderlust/graph.clj
(ns wanderlust.graph
  (:require
   [wanderlust.geometry :as geometry]
   [loom.graph :as graph]
   [loom.alg :as alg]))

(graph/edges (alg/prim-mst (graph/weighted-graph
                            {:a {:b 1 :c 1}
                             :b {:d 1}
                             :c {:d 1 :e 1 :f 1}
                             :d {:f 1}
                             :f {:e 1}})))


(defn pathways-location-graph [pathways]
  (apply graph/weighted-graph
         (map #(vec (:locations %)
                    (:weight %)) pathways)))

(defn pathways-coord-graph [pathways]
  )


(defn mst-graph [coords]
  (->> coords
       (map #(conj % (* (geometry/line-length %) (inc (rand-int 5)))))
       vec
       (apply graph/weighted-graph)
       (alg/prim-mst)
       graph/edges))

(defn pathways-tree [pathways]
  (let [g (pathways-graph pathways)]
    (graph/edges (alg/prim-mst g))))


(comment
  (graph/weighted-graph
      [:a :b 2] [:b :c 3] [:a :c 4])
  )



src/wanderlust/geometry.clj
(ns wanderlust.geometry
  (:require [delaunay-triangulation.core :as delaunay]))

(defn triangulate [points]
  (:edges (delaunay/triangulate points)))

(defn line-length
  [[[x1 y1] [x2 y2]]]
  (Math/hypot (- x1 x2)
              (- y1 y2)))

(defn polyline-length
  [polyline]
  (let [segments (partition 2 1 polyline)]
    (reduce + (map line-length segments))))

(defn point-at-line
  [dist line]
  (let [ratio (/ dist (line-length line))
        [[x1 y1] [x2 y2]] line]
    (when (<= ratio 1)
      [(+ (* x1 (- 1 ratio)) (* x2 ratio))
       (+ (* y1 (- 1 ratio)) (* y2 ratio))])))

(defn polyline->segments
  [polyline]
  (partition 2 1 polyline))

(defn point-at-polyline
  [len polyline]
  (let [segments (polyline->segments polyline)
        slen (->> segments
                  (map line-length)
                  (interleave segments)
                  (partition-all 2))]
    (loop [total-len 0
           ss slen]
      (if (empty? ss)
        (do (prn total-len)
            nil)
        (let [[pts l] (first ss)]
          (if (>= (+ total-len l) len)
            (point-at-line (- len total-len) pts)
            (recur (+ total-len l)
                   (rest ss))))))))

;; http://increasinglyfunctional.com/2013/12/08/point-polygon-clojure/
(defn- crossing-number
  "Determine crossing number for given point and segment of a polygon.
   See http://geomalgorithms.com/a03-_inclusion.html"
  [[px py] [[x1 y1] [x2 y2]]]
  (if (or (and (<= y1 py) (> y2 py))
          (and (> y1 py) (<= y2 py)))
    (let [vt (/ (- py y1) (- y2 y1))]
      (if (< px (+ x1 (* vt (- x2 x1))))
        1 0))
    0))

(defn inside-polygon?
  "Is point inside the given polygon?"
  [point polygon]
  (odd? (reduce +
                (for [n (range (- (count polygon) 1))]
                    (crossing-number point [(nth polygon n)
                                            (nth polygon (+ n
                                                            1))])))))

(defn polygon-bounding [polygon]
  (let [xs (map first polygon)
        ys (map second polygon)]
    [[(apply min xs)
      (apply min ys)]
     [(apply max xs)
      (apply max ys)]]))

(defn- sign [[x1 y1] [x2 y2] [x3 y3]]
  (- (* (- x1 x3) (- y2 y3))
     (* (- x2 x3) (- y1 y3))))

(defn inside-triangle? [p [v1 v2 v3]]
  (let [b1 (< (sign p v1 v2) 0)
        b2 (< (sign p v2 v3) 0)
        b3 (< (sign p v3 v1) 0)]
    (and (= b1 b2) (= b2 b3))))


(defn inside-circle? [point center radius]
  (> radius (line-length [point center])))

(defn polygon->triangles [polygon]
  (map #(conj % (first polygon)) (partition 2 1 (rest polygon))))

(defn inside-convex-polygon? [point polygon]
  (boolean (some #(inside-triangle? point %) (polygon->triangles polygon))))


(comment

  (line-length [[10 40] [11 30]])

  
  (polyline-length [[0 0] [10 10] [20 0] [20 10]])

  
  (point-at-line 5 [[0 0] [10 10]])
  
  (point-at-polyline 15 [[0 0] [10 10] [30 -10] [40 0] [50 0]])


  (inside-polygon? [60 -20] [[10 20] [20 20] [40 110] [100 30] [80 -50]])

  (polygon-bounding [[10 20] [20 20] [40 110] [100 30] [80 -50]])

  
  (inside-convex-polygon? [0 0] [[-10 0] [10 10] [20 0] [20 -10] [10 -25] [0 -30]])

  (inside-polygon? [0 0] [[-10 0] [10 10] [20 0] [20 -10] [10 -25] [0 -30]])

  (polygon->triangles [[-10 0] [10 10] [20 0] [20 -10] [10 -25] [0 -30]])
  )


src/wanderlust/svg_draft.clj
(ns wanderlust.svg-draft
  (:require
   [wanderlust.geometry :as geometry]
   [wanderlust.graph :as graph]
   [dali.io :as io]
   [dali.syntax :as s]
   [clj-xpath.core :refer :all]))


(def ^:private ids (atom {}))

(defn reset-ids! []
  (reset! ids {}))
(defn gen-id!
  ([prefix]
    (gen-id! prefix "-"))
  ([prefix div]
   (format "%s%s%05d"
           prefix div
           (if (get @ids prefix)
             (get (swap! ids update prefix inc) prefix)
             (get (swap! ids assoc prefix 0) prefix)))))


(defn locations-grid [w h s r]
  (for [x (range s w s)
        y (range s h s)]
    (let [opx (rand-nth [+ -])
          opy (rand-nth [+ -])
          rx (rand-int r)
          ry (rand-int r)]
      [(opx x rx) (opy y ry)])))


(defn generate-draft
  ([] (generate-draft {}))
  ([{:keys [name desc width height location-step location-deviation]
     :or   {name "Terra Incognita"
            desc "HC SVNT DRACONES"
            width  1000
            height 500
            location-step 50
            location-deviation 35}}]
   (reset-ids!)
   (let [locations (locations-grid width height location-step location-deviation) ]
     [:dali/page {:width width :height height}
      (s/css "circle.location {stroke: yellow; fill: red;}")
      #_(into [:g {:id "terrain" :inkscape:groupmode "layer" :inkscape:label "terrain"}]
              (voronoi-lines (locations-grid 1000 500 20 18)))
      [:g {:id "terrain shapes" :inkscape:groupmode "layer" :inkscape:label "terrain-shapes"}
       [:rect {:id "primal-ocean" :stroke :midnightblue :fill :midnightblue}
        [0 0] [width height]]]
      [:g {:id "connections" :inkscape:groupmode "layer" :inkscape:label "connections"}
       [:circle {:stroke :black :stroke-width 1 :fill :indigo} [0 0] 1]]
      [:g {:id "pathways" :inkscape:groupmode "layer" :inkscape:label "pathways"}
       [:circle {:stroke :black :stroke-width 1 :fill :indigo} [0 0] 1]]
      (into
        [:g {:id "locations" :inkscape:groupmode "layer" :inkscape:label "locations"}]
        (map (fn [coord] [:circle {:id (gen-id! "location") :class "location"} coord 3]) locations))
      [:g {:id "labels" :inkscape:groupmode "layer" :inkscape:label "labels"}
       [:circle {:stroke :black :stroke-width 1 :fill :indigo} [0 0] 1]]])))


(defn gen-locations [{:keys [width height] :as chart}]
  (let [{:keys [location-step location-deviation]} (:gen-params chart)
        location-points (locations-grid width height location-step location-deviation)
        locations (into {}
                        (map (fn [pt] [(gen-id! "location")
                                       {:coords pt}]) location-points))]
    (assoc chart :locations locations)))



(defn connect-locations [locations pathway]
  (let [[p1 p2] (:coords pathway)
        l1 (some
               (fn [[id {:keys [coords]}]]
                 (when (geometry/inside-circle? p1 coords 3)
                   id))
               locations)
        l2 (some
               (fn [[id {:keys [coords]}]]
                 (when (geometry/inside-circle? p2 coords 3)
                   id))
               locations)]
    (assoc pathway :locations (if (and l1 l2)
                                [l1 l2]
                                :lost))))


(comment
  (connect-locations [["loc1" {:coords [1 1]}]
                      ["loc2" {:coords [10 10]}]
                      ["loc3" {:coords [20 20]}]]
                     {:coords [[20 20] [10 10]]})
  (geometry/inside-circle? [10 10] [10 11] 2)
  )



(defn save-loops [])

(defn discard-long [pathways])

(defn gen-pathways [{:keys [locations] :as chart}]
  (let [weight (get-in chart [:gen-params :pathway-weight-range])
        roads (->> locations
                   vals
                   (map :coords)
                   geometry/triangulate
                   graph/mst-graph              
                   (map set)
                   distinct
                   (map vec)
                   (map (fn [c] {:coords c
                                 :weight (inc (rand-int weight))
                                 :length (geometry/line-length c)}))
                   (map (partial connect-locations locations)))]
    (assoc chart :draft-pathways roads)))


(defn generate-chart
  ([] (generate-chart {} {}))
  ([map-params gen-params]
   (reset-ids!)
   (let [default-map
         {:name "Terra Incognita"
          :desc "HC SVNT DRACONES"
          :width  1000
          :height 500
          :gen-params {:location-step 50
                       :location-deviation 35
                       :pathway-weight-range 5}
          :locations {}
          :pathways {}
          :terrain {}
          }
         chart (-> default-map
                   (merge map-params)
                   (update :gen-params merge gen-params))]
     (-> chart
         gen-locations
         gen-pathways
         ))))

(comment
  (let [chart (generate-chart {:width 1000 :height 1000} {:location-step 80 :location-deviation 50})
       draft (chart->svg-draft chart)
        ]
    chart
    (io/render-svg draft "test.svg")
    )
)

(defn svg-draft->chart [svg-draft]
  (let [svg-doc (xml->doc svg-draft)]
    (prn svg-doc)
    (let [locations
          (reduce
            (fn [res {:keys [attrs text]}]
              (prn attrs text)
              (assoc res (:id attrs)
                         {:coords [(Double/parseDouble (:cx attrs))
                                   (Double/parseDouble (:cy attrs))]
                          :data   (clojure.edn/read-string text)}))
            {} ($x "//svg/g[@id='locations']/circle" svg-doc))]
      {:locations locations})))

(comment
  (svg-draft->chart (slurp "./test.svg"))
  )

(defn location-draft [[id {:keys [coords]}]]
  [:circle {:id id
            :stroke :yellow
            :fill :red}
   coords
   3])

(defn pathways-draft [{:keys [coords weight length]}]
  (let [[p1 p2] coords
        color (case weight
                1 :blue
                2 :darkmagenta
                3 :darkred
                4 :red
                5 :yellow)]
    [:line {:stroke color
            :stroke-width 1}  p1 p2]))

;(pathways-draft {:coords [[1 2] [3 4]] :weight 3 :length 10.1})

(defn chart->svg-draft [chart]
  (let [{:keys [name desc width height locations draft-pathways]} chart]
    [:dali/page {:width width :height height}
     [:title name] [:desc desc]
     [:g {:id "terrain shapes"
          :inkscape:groupmode "layer"
          :inkscape:label "terrain-shapes"}]
     [:g {:id "terrain shapes"
          :inkscape:groupmode "layer"
          :inkscape:label "terrain-shapes"}]
     [:g {:id "terrain shapes"
          :inkscape:groupmode "layer"
          :inkscape:label "terrain-shapes"}]
     (into [:g {:id "locations"
                :inkscape:groupmode "layer"
                :inkscape:label "locations"}]
           (map location-draft locations))
     (into [:g {:id "pathways"
                :inkscape:groupmode "layer"
                :inkscape:label "pathways"}]
           (map pathways-draft draft-pathways))
     ]))

(comment
  (io/render-svg
    (generate-draft {:width 1000 :height 1000 :location-step 35 :location-deviation 30})
    "test.svg"))
